test/instruction_graph_test_utils.h

```diff
@@ -0,0 +1,736 @@
+	static std::string print_filter(const Predicate& /* predicate */) {
+		return "<lambda>";
+	}
+
+	template <typename GeneralRecord, typename SpecificRecord, typename... Filters>
+	std::string filter_trace(const std::string& selector, const Filters&... filters) const {
+		auto trace = m_trace.empty() ? selector : fmt::format("{}.{}", m_trace, selector);
+		if constexpr(!std::is_same_v<GeneralRecord, SpecificRecord>) { fmt::format_to(std::back_inserter(trace), "<{}>", kernel_debug_name<SpecificRecord>()); }
+
+		std::string param_trace;
+		// [[maybe_unused]]: GCC will emit an unused-warning if `Filters...` is empty and the comma-fold expression below expands to nothing
+		[[maybe_unused]] const auto format_filter_param = [&](const auto& f) {
+			if(!param_trace.empty()) { param_trace += ", "; }
+			param_trace += instruction_query<SpecificRecord>::print_filter(f);
+		};
+		(format_filter_param(filters), ...);
+		fmt::format_to(std::back_inserter(trace), "({})", param_trace);
+		return trace;
+	}
+
+	template <typename... InstructionQueries>
+	instruction_query<> union_with(const InstructionQueries&... tail) const {
+		assert(((m_recorder == tail.m_recorder) && ...));
+
+		// construct union in recorder-ordering without duplicates - always returns the base type
+		std::vector<const instruction_record*> union_query;
+		for(const auto& instr : m_recorder->get_instructions()) {
+			if(std::find(m_result.begin(), m_result.end(), instr.get()) != m_result.end()
+			    || ((std::find(tail.m_result.begin(), tail.m_result.end(), instr.get()) != tail.m_result.end()) || ...)) { //
+				union_query.push_back(instr.get());
+			}
+		}
+
+		std::string trace = fmt::format("union_of({}", m_trace);
+		(((trace += ", ") += tail.m_trace), ...);
+		trace += ")";
+		return instruction_query<>(m_recorder, std::move(union_query), std::move(trace));
+	}
+
+	template <typename... InstructionQueries>
+	instruction_query intersection_with(const InstructionQueries&... tail) const {
+		assert(((m_recorder == tail.m_recorder) && ...));
+
+		// construct intersection in recorder-ordering without duplicates - returns the type of `head`
+		std::vector<const Record*> intersection_query;
+		for(const auto& instr : m_result) {
+			if(((std::find(tail.m_result.begin(), tail.m_result.end(), instr) != tail.m_result.end()) && ...)) { //
+				intersection_query.push_back(instr);
+			}
+		}
+
+		std::string trace = fmt::format("intersection_of({}", m_trace);
+		(((trace += ", ") += tail.m_trace), ...);
+		trace += ")";
+		return instruction_query(m_recorder, std::move(intersection_query), std::move(trace));
+	}
+
+	instruction_query(const instruction_recorder* recorder, std::vector<const Record*> query, std::string trace)
+	    : m_recorder(recorder), m_result(std::move(query)), m_trace(std::move(trace)) {}
+};
+
+/// A view on an `instruction_recorder` like `instruction_query`, but for `outbound_pilot`s.
+class pilot_query {
```

Comment:
Pilots are transmitted immediately upon being received by the executor thread, so they don't have the need for dependency tracking (which would qualify them as real DAG nodes).

We did discuss `pilot_instructions` (in person) that would batch up these pilots to delay sending them somewhat by tieing them to epochs in order to avoid clogging the interconnect in pilot messages if that ever becomes a concern (similar to the MPI throughput issues we observed in the old master-worker model).

Reply:
I thought that this was the reasoning, but what I was wondering is whether -- even if they have no dependencies (for now) -- it would make sense to treat them as instructions. There's quite a lot of code (I'd say several hundered lines?) that only exists because they aren't. Would there be any drawback?

Reply:
I deliberately want to make an effort here to keep everything out of the hot executor loop that does not need dependency tracking. Also adding instructions without dependencies would violate the invariant that every instruction transitively depends on an epoch or horizon. The additional code is mostly test code I believe